### DISCUSSION TITLE
Resolving fields of root queries

### DISCUSSION THREAD

**Discussion by erquhart** [urql-graphql/urql#3100] [Q&A]
I've found I have to use `query_root` for my root field resolvers to work (possibly due to using Hasura?). So my resolvers config looks like this:

```ts
    resolvers: {
      query_root: {
        account: (parent, args) => {
          return handleQueryById('Account', args)
        },
```

I'd like to transform individual fields of `account`, specifically turning `null` to `undefined` in this case, but it doesn't seem possible since the resolvers config above makes `account` the field and it doesn't seem I can resolve sub-fields.

I'm also wondering, if this is indeed possible, does the integration between graphql-codegen and urql take the resolvers config into account? Or is anyone aware of any examples of leveraging the urql config to get more accurate types out of graphql-codegen?

Appreciate any help you can give, been banging my head on this for a while.

**Comment by kitten** [comment]
If you need to transform fields on `Account` then you may want to create field resolvers on that type instead.

> I'd like to transform individual fields of account, specifically turning null to undefined in this case

That's not possible. Both values have a special meaning in GraphQL results, and hence resolvers treat them in a special way. Specifically, `undefined` is treated as an uncached fields.

I'm obviously not sure what you're trying to do, so it'd be pretty hard to elaborate on any possible solutions 😅

**Reply by erquhart** [reply]
> If you need to transform fields on Account then you may want to create field resolvers on that type instead.

I'll try that, I landed on `query_root` because resolvers on `resolvers.account.[field]` weren't having any affect.

For `null`/`undefined`, Hasura returns null for empty fields, which is expected because the value is indeed null, but for all intents and purposes in the client, `NULL` in SQL is ideally expressed as `undefined` in the client. I want functions to know the field has no value, literally "undefined". And I'm okay with urql treating these fields as uncached.

Currently I'm using a normalization function to transform both values and their types from null to undefined, recursively, anywhere in the application where I'm working with data. Doable, but a lot of boilerplate that it'd be nice to do away with.

**Reply by kitten** [reply]
> NULL in SQL is ideally expressed as undefined in the client.

That's not quite the full story unfortunately 😅 While it's an often repeated statement that "JavaScript has two NULLs" that's not quite the full story, so I get the sentiment, but there's two important notes to add to this:
- `null` is an actually serialisable value and represented in JSON (so `null` is serialised, `undefined` is left out)
- `undefined` is different from `null`, as per the above, since `null` is a NULL-ish value, while `undefined` stands for the lack of a value, i.e. the opposite of a value being present.
  - (Sidenote: Of course we could go deeper here, since unfortunately `undefined` is a declarable identifier, but oh well)

The short story is, GraphQL has made some decisions for us here, which are pretty sensible, since it adheres to JSON serialization (and hence to JS, since JS’ scalars are basically JSON scalars).
This is captured here in the specification: https://spec.graphql.org/October2021/#sec-JSON-Serialization

The other part of this story is that Graphcache actually uses `undefined` do identify the lack of a value. Internally, for our own logic, we of course didn't have to do this, but for the resolvers it's actually important, since otherwise, there isn't an intuitive way to actually write a resolver that says "there isn't a value for this", since `null` _is_ in fact a value.

If you do want to change this, you're better off writing a mapping exchange (simplest being, with the `mapExchange`) that replaces `null` with `undefined`.

I'd really not recommend this however. This will get particularly awkward once `@defer` becomes popular and you're looking to use this, as deferred values will be "left out" (i.e. they'll be `undefined`), which will again be different from `null` (i.e. `null` values from the API)

This is a hard problem in schema design, since in some cases you'll make decisions that causes your GraphQL schema to maybe have different meanings for nullable values (it could be the lack of a value, or something else 🤷)

Either way, from our perspective, and we've said this before, Hasura isn't “optimal” in making some of this obvious, including some details of schema design and its importance in GraphQL.

But I hope this makes it clearer why Graphcache’s resolvers aren't the right place to do this, and why — potentially — it's probably not a good idea to bother at all 😅 

Anyway, tl;dr Hasura doesn't return `null` for fields, but GraphQL schemas may be modeled to have nullable fields for missing DB entries/items (as a choice), rather than a union-object value, or an error — which of course are all valid methods of conveying the lack of a value — but the GraphQL spec itself and JSON serialization mean that `null` is a preferred value for a “present” value as delivered by a GraphQL API.
